tests/mail/tests.py

```diff
@@ -556,12 +558,50 @@ def test_alternatives(self):
         mime_type = "text/html"
         msg.attach_alternative(html_content, mime_type)
 
+        self.assertIsInstance(msg.alternatives[0], EmailAlternative)
+
+        self.assertEqual(msg.alternatives[0][0], html_content)
+        self.assertEqual(msg.alternatives[0].content, html_content)
+
+        self.assertEqual(msg.alternatives[0][1], mime_type)
+        self.assertEqual(msg.alternatives[0].mimetype, mime_type)
+
+        msg.message()
```

Comment:
Each of the tests end with `msg.message()` without further asserts, are you wanting to check this doesn't fail? Can we either remove or assert something?

Reply:
Sorry, that was my request from [here](https://github.com/django/django/pull/18318#discussion_r1657696314)

The goal is to verify we maintain compatibility with using _unnamed_ tuples in EmailMessage.attachments (which was documented and hasn't been deprecated):

```python
msg = django.core.mail.EmailMessage(...)
msg.attachments = [("attachment.txt", "this is attached", "text/plain")]
msg.send()
```

Checking that msg.message() doesn't fail checks that someone hasn't tried to simplify the EmailMessage._create_attachment… internals by using the tuple attr names.

Something like
```python
with self.assertNotRaises(): 
    msg.message()
``` 
would be appropriate, but assertNotRaises doesn't exist.

Maybe this?
```python
self.assertIsNotNone(msg.message())  # assert doesn't raise
```

Or this?
```python
try:
    msg.message()
except AttributeError:
    self.fail("Email.attachments must support unnamed tuples")
```

Even better would be checking that the expected attachment actually made it into the result of msg.message(), but that feels out of scope for this change. (And will be much easier once that result can use Python's modern [`EmailMessage.iter_attachments()`](https://docs.python.org/3/library/email.message.html#email.message.EmailMessage.iter_attachments). I'm planning those test updates as part of ticket-35581.)

Reply:
Looking at some of the other tests in the file, maybe this one something like
```python
self.assertIn(html_content, msg.message().as_string())
```
and in the ones for attachments
```python
payload = msg.message().get_payload()
self.assertEqual(payload[0].get_filename(), file_name)
```
No reason to say sorry! Also good suggestions 👍

Reply:
Those assertions on the message content/payload seem ideal.

Reply:
For attachments, there's also a MailTests.get_decoded_attachments() helper:

```python
attachments = self.get_decoded_attachments(msg)
self.assertEqual(attachments[0], (filename, content, mimetype))
```

get_decoded_attachments() serializes msg.message() and parses the result, so it's a gold-standard verification the attachments make it through correctly. (You might even update it to return an EmailAttachment namedtuple :smile:)

Reply:
Ooh, that's even nicer!